File: Metode Avansate de Programare/C#/Laborator12-13/Laborator12-13/Repository/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Laborator12_13.domain;
using Laborator12_13.Validator;
using Laborator12_13.repository;

namespace Laborator12_13.repository
{
    public class InMemoryRepository<ID, E> : ICrudRepository<ID, E> where E : IHasID<ID>
    {

        public IDictionary<ID, E> entities = new Dictionary<ID, E>();
        private IValidator<E> validator;

        public InMemoryRepository(IValidator<E> validator)
        {
            this.validator = validator;
        }

        public E delete(ID id)
        {
            if (id == null)
                throw new ArgumentNullException("Parametrul este null");
            if (!entities.ContainsKey(id))
                return default(E);
            else
            {
                E temp = entities[id];
                entities.Remove(id);
                return temp;
            }
        }

        public IEnumerable<E> findAll()
        {
            return entities.Values.ToList();
        }

        public E findOne(ID id)
        {
            if (id == null)
                throw new ArgumentNullException("Parametrul este null");
            if (!entities.ContainsKey(id))
                return default(E);
            else return entities[id];
        }

        public E save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("Parametrul este null");
            validator.Validate(entity);
            if (entities.ContainsKey(entity.ID))
                return entity;
            entities[entity.ID] = entity;
            return default(E);
        }

        public E update(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("Parametrul este null");
            validator.Validate(entity);
            if (!entities.ContainsKey(entity.ID))
                return entity;
            entities[entity.ID] = entity;
            return default(E);
        }
    }
}

File: Proiect/src/Repository/InMemoryRepository.java
package Repository;

import Domain.HasID;
import Domain.Validare.Validator;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends HasID<ID>> implements CrudRepository<ID,E> {

    private Validator<E> validator;
    Map<ID,E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities=new HashMap<ID,E>();
    }

    /**
     * Returneaza un obiect de tip E ce are ID-ul dat
     * @param id - id-ul obiectului cautat
     * @return obiectul cautat daca acesta exista
     * @throws IllegalArgumentException daca id-ul dat este null
     */
    @Override
    public E findOne(ID id){
        if (id==null)
            throw new IllegalArgumentException("id must be not null");
        return entities.get(id);
    }

    /**
     * Returneaza un iterabil cu toate valorile salvate
     * @return interabil cu entiteati de tip E
     */
    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if (entity == null) throw new IllegalArgumentException();
        validator.validate(entity);
        if (entities.get(entity.getID()) == null) {
            entities.put(entity.getID(), entity);
            return null;
        }
        return entity;
    }

    @Override
    public E delete(ID id) {
        if(id==null) {
            throw new IllegalArgumentException("Parametrul dat este null");
        }
        else{
            E s=findOne(id);
            if(s==null) return null;
            else{
                entities.remove(id);
                return s;
            }
        }
    }

    @Override
    public E update(E entity) {

        if(entity == null)
            throw new IllegalArgumentException("entity must be not null!");
        validator.validate(entity);
        if(entities.get(entity.getID()) != null) {
            entities.put(entity.getID(),entity);
            return null;
        }
        return entity;

    }

}
